File: 3739-count-subarrays-with-majority-element-ii/3739-count-subarrays-with-majority-element-ii.java
class Solution {
    public long countMajoritySubarrays(int[] nums, int target) {
        int n = nums.length;
        long cnt = 0;
        for (int i = 0; i < n; i++) {
            if (nums[i] == target) nums[i] = 1;
            else nums[i] = -1;
        }
        int[] pref = new int[n];
        pref[0] = nums[0];
        for (int i = 1; i < n; i++) {
            pref[i] = pref[i - 1] + nums[i];
        }
        int shift = n;
        int[] freq = new int[2 * n + 1];
        freq[shift] = 1;
        long valid = 0;
        int lastSum = 0;
        for (int i = 0; i < n; i++) {
            if (pref[i] > lastSum) {
                valid += freq[lastSum + shift];
            } else {
                valid -= freq[pref[i] + shift];
            }
            cnt += valid;
            freq[pref[i] + shift]++;
            lastSum = pref[i];
        }
        return cnt;
    }
}

File: 3739-count-subarrays-with-majority-element-ii/3739-count-subarrays-with-majority-element-ii.py
class Solution:
    def countMajoritySubarrays(self, nums, target):
        n = len(nums)
        cnt = 0
        for i in range(n):
            if nums[i] == target:
                nums[i] = 1
            else:
                nums[i] = -1
        pref = [0] * n
        pref[0] = nums[0]
        for i in range(1, n):
            pref[i] = pref[i - 1] + nums[i]
        shift = n
        freq = [0] * (2 * n + 1)
        freq[shift] = 1
        valid = 0
        lastSum = 0
        for i in range(n):
            if pref[i] > lastSum:
                valid += freq[lastSum + shift]
            else:
                valid -= freq[pref[i] + shift]
            cnt += valid
            freq[pref[i] + shift] += 1
            lastSum = pref[i]

        return cnt